### PR TITLE
Fix Gradient2D Repeat repeat mode

### PR DESCRIPTION
**Pull request by MewPurPur**
Fixes this:

image

### CODE REVIEW

**Comment by YuriSizov** [comment]
> Fixes this

And what exactly is "this"? :) Can you provide a before and after at least?

**Comment by MewPurPur** [comment]
Omg sorry, I thought it was obvious! The dark corners aren't supposed to be there, because the "TO" handle is in the very corner.

**Comment by kleonc** [comment]
It doesn't fix the issue, it just moves it into another case:

|v4.0.2.stable.official [7a0977ce2]|v4.1.dev.custom_build [e25737d72]|
|-|-|
|Godot_v4 0 2-stable_win64_bdMnQJGiLK|rq1W3thhl1|

**Comment by MewPurPur** [comment]
Aw man, back to the drawing board then.

**Comment by MewPurPur** [comment]
I think the correct solution would be to make it so:
- The offsets mapped to 0 are: `-2`, `-1`, `0`, `1 + ε`, `2 + ε`
- The ranges mapped to 1 are: `-1 - ε`, `-ε`, `1`, `2`, `3`

**Comment by RedworkDE** [comment]
(Not at my PC to test any of this, so this is just a guess from looking at this one function)

What happens when the gradient goes from 0.5,0 to 1,0? Probably 0,0 is the wrong color still.

I think the original repeat code is correct (It actually never maps any point to the offset 1.0 of the gradient) and the standard way of doing it.

Instead I think the issue is way the sampling position is selected for each pixel.
Currently the top left pixel is sampled in the top left corner, the bottom righ pixel at the bottom right corner.

This means (and is presumably why it was done like it in the first place) that for a simple gradient across the whole texture, the edges of the of the textures are the exact color of the edgde of the gradient.

But I think this behavior is wrong, instead the pixel should be the average color of the gradient across the area of the pixel, which for smooth gradients can be cheaply approximated by sampling each pixel in its center.

Basically increasing the resolution of a gradient texture and then downsampling it again should result in the same image as rendering it at that resolution directly.

TL/DR: Instead of this change the two ifs checking for size > 1 should be replaced by `pos = pixel + 0.5 / size`

**Comment by MewPurPur** [comment]
> What happens when the gradient goes from 0.5,0 to 1,0? Probably 0,0 is the wrong color still.

I think it's good:

image

> Basically increasing the resolution of a gradient texture and then downsampling it again should result in the same image as rendering it at that resolution directly.

Take kleonc's gradient. I think users expect this:

image

If I offset half a pixel here, I would for example not get black on the leftmost pixel, but some dark gray, wouldn't I?

**Comment by kleonc** [comment]
> I think the correct solution would be to make it so:
> 
>     * The offsets mapped to 0 are: `-2`, `-1`, `0`, `1 + ε`, `2 + ε`
> 
>     * The offsets mapped to 1 are: `-1 - ε`, `-ε`, `1`, `2`, `3`

I don't think that's a correct solution. For confirmation:

v4.1.dev.custom_build [59d67988e]:
B7lJdM7s9FcMf6X9pp80godot windows editor x86_64_gvdrbOfw0C

---

To me the issue seems more conceptual, not just a simple calculation bug. Gradient conceptually is continous, texture is not. GradientTexture2D is a discrete set of the gradient's samples. The issue seems to be how this sampling is done / defining what is the expected result, I agree with @RedworkDE on that. Not sure what's the expected behavior/solution.

**Comment by MewPurPur** [comment]
Hmm, okay, maybe this doesn't fix everything... Still, it maps 0 to 0 and 1 to 1, so I think it's a W compared to before nonetheless?
